packages/cli/src/lib/push/parseFile.ts

```diff
@@ -26,17 +32,28 @@ export const parseAndGetDocumentBuilderFromJSONDocument = (
 };
 
 const processDocument = (
-  fileContent: Record<string, string>,
+  fileContent: Record<string, string | {}>,
   documentPath: PathLike
 ) => {
   const caseInsensitiveDoc = new CaseInsensitiveDocument(fileContent);
 
   const documentBuilder = validateRequiredKeysAndGetDocumentBuilder(
-    caseInsensitiveDoc,
+    caseInsensitiveDoc as CaseInsensitiveDocument<string>,
     documentPath
   );
-  processKnownKeys(caseInsensitiveDoc, documentBuilder);
-  processMetadata(caseInsensitiveDoc, documentBuilder);
+  processKnownKeys(
+    caseInsensitiveDoc as CaseInsensitiveDocument<string>,
+    documentBuilder
+  );
+  processSecurityIdentities(
+    caseInsensitiveDoc as CaseInsensitiveDocument<Document['permissions']>,
+    documentBuilder,
+    documentPath
+  );
+  processMetadata(
+    caseInsensitiveDoc as CaseInsensitiveDocument<MetadataValue>,
+    documentBuilder
+  );
```

Comment:
I think there's really too much casting.

That smells like a logic that reaches its ends and that you fight against it.

Maybe it's time to rethink this?

Reply:
I'll make use of the `PrimitiveValues` type from Bueno a bit more.
Still, there's some casting that needs to happen since, inherently, all values in an arbitrary JSON documents are "unknown" upfront.